src/compiler/interpreter.jl

```diff
@@ -0,0 +1,112 @@
+        @assert world <= Base.get_world_counter()
+
+        return new(
+            cache,
+            mt,
+
+            # Initially empty cache
+            Vector{InferenceResult}(),
+
+            # world age counter
+            world,
+
+            # parameters for inference and optimization
+            InferenceParams(unoptimize_throw_blocks=false),
+            VERSION >= v"1.8.0-DEV.486" ? OptimizationParams() :
+                                          OptimizationParams(unoptimize_throw_blocks=false),
+        )
+    end
+end
+
+Core.Compiler.InferenceParams(interp::EnzymeInterpeter) = interp.inf_params
+Core.Compiler.OptimizationParams(interp::EnzymeInterpeter) = interp.opt_params
+Core.Compiler.get_world_counter(interp::EnzymeInterpeter) = interp.world
+Core.Compiler.get_inference_cache(interp::EnzymeInterpeter) = interp.local_cache
+Core.Compiler.code_cache(interp::EnzymeInterpeter) = WorldView(interp.global_cache, interp.world)
+
+# No need to do any locking since we're not putting our results into the runtime cache
+Core.Compiler.lock_mi_inference(interp::EnzymeInterpeter, mi::MethodInstance) = nothing
+Core.Compiler.unlock_mi_inference(interp::EnzymeInterpeter, mi::MethodInstance) = nothing
+
+function Core.Compiler.add_remark!(interp::EnzymeInterpeter, sv::InferenceState, msg)
+end
+
+Core.Compiler.may_optimize(interp::EnzymeInterpeter) = true
+Core.Compiler.may_compress(interp::EnzymeInterpeter) = true
+Core.Compiler.may_discard_trees(interp::EnzymeInterpeter) = true
+if VERSION >= v"1.7.0-DEV.577"
+Core.Compiler.verbose_stmt_info(interp::EnzymeInterpeter) = false
+end
+
+if isdefined(Base.Experimental, Symbol("@overlay"))
+Core.Compiler.method_table(interp::EnzymeInterpeter, sv::InferenceState) =
+    Core.Compiler.OverlayMethodTable(interp.world, interp.method_table)
+else
+Core.Compiler.method_table(interp::EnzymeInterpeter, sv::InferenceState) =
+    GPUCompiler.WorldOverlayMethodTable(interp.world)
+end
+
+function is_primitive_func(@nospecialize(TT))
+    isa(TT, DataType) || return false
+    ft = TT.parameters[1]
+    if ft === typeof(Base.cbrt) || ft === typeof(Base.sin) || ft === typeof(Base.cos) ||
+       ft === typeof(Base.tan) || ft === typeof(Base.exp) || ft === typeof(Base.log) ||
+       ft === typeof(Base.asin) || ft === typeof(Base.tanh) || ft === typeof(Base.FastMath.tanh_fast)
+
+        if TT <: Tuple{ft, Float32} || TT <: Tuple{ft, Float64}
+            return true
+        end
+    end
+    return false
+end
+
+
+# branch on https://github.com/JuliaLang/julia/pull/41328
+@static if isdefined(Core.Compiler, :is_stmt_inline)
+
+function Core.Compiler.inlining_policy(
+    interp::EnzymeInterpeter, @nospecialize(src), stmt_flag::UInt8,
+    mi::MethodInstance, argtypes::Vector{Any})
+
+    if is_primitive_func(mi.specTypes)
+        return nothing
+    end
+
+    return Base.@invoke Core.Compiler.inlining_policy(
+        interp::AbstractInterpreter, nothing, stmt_flag::UInt8,
```

Comment:
```suggestion
        interp::AbstractInterpreter, src::Any, stmt_flag::UInt8,
```
`src === nothing` here means we may inline only those calls with callsite `@inline` annotations.